### ISSUE TITLE
Replicating Caffe performance?

### ISSUE THREAD

**Issue by Yuliang-Zou** [ildoonet/tf-pose-estimation#218]
Hi @ildoonet , thanks for sharing the nice code!

I wonder how you can replicate the numbers by training from scratch. I load vgg pretrained, lr=1e-4, batch_size=16, I trained for ~140k iterations and found the training loss basically converged. Then I test it on val2014, but I only got AP=0.3360.

Can you share your settings?

**Comment by Yuliang-Zou** [comment]
Not really

**Comment by mgarbade** [comment]
Replicating the Caffe performantce seems to be a long road ahead, so I try to list some of the problems I could find so far, and keep updating the list as I move along.

Here are some of the differences between the original caffe code and the code in this repository:

First some links to the original training code for vgg19 (first 10 layers) as backbone reaching  58.80%  on coco val 2017 (multiscale evaluation) and 48.99 % mAP (single scale evaluation)

[pose_train_test.prototxt](https://github.com/ZheC/Realtime_Multi-Person_Pose_Estimation/blob/master/training/example_proto/pose_train_test.prototxt), shows the original architecture, as well as preprocessing steps, learning rate multipliers, gaussian weight initializers and weight decay parameters

[pose_solver.prototxt](https://github.com/ZheC/Realtime_Multi-Person_Pose_Estimation/blob/master/training/example_proto/pose_solver.prototxt), specifies optimizer and lr policy (default optimizer in Caffe is SGD with momentum)


  | TFPE (this repo) | CMU (original caffe version)
-- | -- | --
preprocessing | flipcrop_xy = 432x368<br>rotate: 15°<br>scale_min: 0.8 <br> scale_max: 1.2| stride: 8    <br>max_rotate_degree: 40   <br>crop_size_x: 368   <br>crop_size_y: 368    <br>scale_prob: 1    <br>scale_min: 0.5    <br>scale_max: 1.1    <br>target_dist: 0.6 <- ???    <br>center_perterb_max: 40 <- ??? <br> ???: probably computed based on some bounding box values 
weight initialization |   | gaussian, stdev=0.01 (conv)constant (bias)
learning rate multiplier | none | 1.0 for base weights<br>2.0 for base bias<br>4.0 for CPM + Mconv weights<br>8.0 for CPM + Mconv weights
optimizer | Adam | SGD + Momentum (0.9)
learning rate | 0.001 | 4e-05
lr policy | cosine decay | step (after 10 Epochs?)
weight decay | none | 5e-4<br>decay_mult_conv = 1<br>decay_mult_bias = 0
batch size | 64? 32! | 10
losses | l2 loss averaged over all stages <br>(only 6 are used!?) using ‘reduce mean’ for both heatmap and paf<br>paf and heatmap loss are averaged<br>the division by batchsize is ‘wrong’ in case of multi-gpu training | only paf loss for all 6 stages<br>heatmap loss in last stage only (since 30.05.2019)
background class | learns a background class:<br>1 - max(heatmap, axis=2)→ total of 19 classes | not mentioned here!?
